api/server/secrets.go

```diff
@@ -0,0 +1,148 @@
+	"net/http"
+
+	"github.com/libopenstorage/openstorage/secrets"
+)
+
+const (
+	secretKeyOkMsg   = "Secret Key set successfully"
+	secretLoginOkMsg = "Secrets Login Succeeded"
+	secretLoginCheck = "Secrets Login Check Succeeded"
+)
+
+// TODO: Add swagger yaml
+func (c *clusterApi) setDefaultSecretKey(w http.ResponseWriter, r *http.Request) {
+
+	method := "setDefaultSecretKey"
+	var secReq secrets.DefaultSecretKeyRequest
+	var secResp secrets.SecretResponse
+
+	if err := json.NewDecoder(r.Body).Decode(&secReq); err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	err := c.SecretManager.SetDefaultSecretKey(
+		secReq.DefaultSecretKey,
+		secReq.Override,
+	)
+
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	secResp.Status = "Default " + secretKeyOkMsg
+	json.NewEncoder(w).Encode(secResp)
+}
+
+// TODO: Add swagger yaml
+func (c *clusterApi) getDefaultSecretKey(w http.ResponseWriter, r *http.Request) {
+
+	method := "getDefaultSecretKey"
+
+	secretValue, err := c.SecretManager.GetDefaultSecretKey()
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	secResp := &secrets.GetSecretResponse{
+		SecretValue: secretValue,
+	}
+	json.NewEncoder(w).Encode(secResp)
+
+}
+
+// TODO: Add swagger yaml
+func (c *clusterApi) secretsLogin(w http.ResponseWriter, r *http.Request) {
+	var secReq secrets.SecretLoginRequest
+	var secResp secrets.SecretResponse
+	method := "secretsLogin"
+
+	if err := json.NewDecoder(r.Body).Decode(&secReq); err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	err := c.SecretManager.Login(secReq.SecretType, secReq.SecretConfig)
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	secResp.Status = secretLoginOkMsg
+	json.NewEncoder(w).Encode(secResp)
+}
+
+// TODO: Add swagger yaml
+func (c *clusterApi) setSecret(w http.ResponseWriter, r *http.Request) {
+
+	method := "setSecret"
+	var secReq secrets.SetSecretRequest
+	var secResp secrets.SecretResponse
+	params := r.URL.Query()
+	secretID := params[secrets.SecretKey]
+
+	if len(secretID) == 0 || secretID[0] == "" {
+		c.sendError(c.name, method, w, "Missing secret ID", http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&secReq); err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	err := c.SecretManager.Set(secretID[0], secReq.SecretValue)
+	if err != nil {
+		c.sendError(c.name, method, w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	secResp.Status = secretKeyOkMsg
+	json.NewEncoder(w).Encode(secResp)
+}
+
+// TODO: Add swagger yaml
+func (c *clusterApi) getSecret(w http.ResponseWriter, r *http.Request) {
+
+	method := "getSecret"
+	params := r.URL.Query()
```

Comment:
same here.  Or here we can use the Gorilla mux to get the value from the URI as a path but not as a query.

For example instead of:

```go
{verb: "GET", path: clusterSecretPath("", cluster.APIVersion), fn: c.getSecret},
{verb: "PUT", path: clusterSecretPath("", cluster.APIVersion), fn: c.setSecret},
```

it could be 

```go
{verb: "GET", path: clusterSecretPath("/{secretKey}", cluster.APIVersion), fn: c.getSecret},
{verb: "PUT", path: clusterSecretPath("/{secretKey}", cluster.APIVersion), fn: c.setSecret},
```

where in `PUT` it passes a JSON object containing the secret value

Reply:
but using mux will contradict with other api's ?
Right now we have API as,
```
GET /cluster/secrets/<secret_id>
PUT /cluster/secrets/<secret_id> Body: {value: }
GET /cluster/secrets/defaultsecretkey
PUT /cluster/secrets/defaultsecretkey Body: {value: }
GET /cluster/secrets/verify
POST /cluster/secrets/login Body: {}
```
so both call to GET /cluster/secrets/myid and GET /cluster/secrets/defaultsecretkey  will ultimately goes to route ```{verb: "GET", path: clusterSecretPath("/{secretKey}", cluster.APIVersion), fn: c.getSecret},```

Am I wrong here?

Reply:
You are correct here. Ignore my request.

Reply:
ok, thanks Luis